File: coord64/julia/kd.jl
module KD

export Tree, Box, insert!, count_items, is_member, hard_delete!, search, LEFT, BOTTOM, RIGHT, TOP

const Box = NTuple{4, Int64}
const LEFT = 1
const BOTTOM = 2
const RIGHT = 3
const TOP = 4

mutable struct Node{T}
    item::Union{T, Nothing}
    size::Box
    lo_min_bound::Int64
    hi_max_bound::Int64
    other_bound::Int64
    sons::Vector{Union{Node{T}, Nothing}}
    
    function Node{T}(item, size::Box, lo::Int, hi::Int, ob::Int) where T
        new{T}(item, size, lo, hi, ob, Union{Node{T}, Nothing}[nothing, nothing])
    end
end

mutable struct Tree{T}
    root::Union{Node{T}, Nothing}
    item_count::Int
    dead_count::Int
    extent::Vector{Int64}
    
    function Tree{T}() where T
        new{T}(nothing, 0, 0, [0, 0, 0, 0])
    end
end

next_disc(disc::Int) = (disc % 4) + 1

function bounds_update!(node::Node{T}, disc::Int, size::Box) where T
    vert = ((disc - 1) & 1) + 1
    node.lo_min_bound = min(node.lo_min_bound, size[vert])
    node.hi_max_bound = max(node.hi_max_bound, size[vert + 2])
    if ((disc - 1) & 2) != 0
        node.other_bound = min(node.other_bound, size[vert])
    else
        node.other_bound = max(node.other_bound, size[vert + 2])
    end
end

function insert!(tree::Tree{T}, item::T, size::Box) where T
    if tree.root === nothing
        tree.root = Node{T}(item, size, size[LEFT], size[RIGHT], size[LEFT])
        tree.extent = [size[1], size[2], size[3], size[4]]
        tree.item_count = 1
        return
    end

    if insert_recursive!(tree.root, 1, item, size)
        tree.item_count += 1
        tree.extent[LEFT] = min(tree.extent[LEFT], size[LEFT])
        tree.extent[BOTTOM] = min(tree.extent[BOTTOM], size[BOTTOM])
        tree.extent[RIGHT] = max(tree.extent[RIGHT], size[RIGHT])
        tree.extent[TOP] = max(tree.extent[TOP], size[TOP])
    end
end

function insert_recursive!(elem::Node{T}, disc::Int, item::T, size::Box) where T
    if elem.item !== nothing && elem.item == item
        return false
    end

    val = size[disc] - elem.size[disc]
    if val == 0
        ndisc = next_disc(disc)
        while ndisc != disc
            val = size[ndisc] - elem.size[ndisc]
            if val != 0
                break
            end
            ndisc = next_disc(ndisc)
        end
        if val == 0
            val = 1
        end
    end

    child_idx = val >= 0 ? 2 : 1

    if elem.sons[child_idx] !== nothing
        inserted = insert_recursive!(elem.sons[child_idx], next_disc(disc), item, size)
        if inserted
            bounds_update!(elem, disc, size)
        end
        return inserted
    end

    vert = ((next_disc(disc) - 1) & 1) + 1
    lo = size[vert]
    hi = size[vert + 2]
    ob = (((next_disc(disc) - 1) & 2) != 0) ? size[vert] : size[vert + 2]

    new_node = Node{T}(item, size, lo, hi, ob)
    elem.sons[child_idx] = new_node
    bounds_update!(elem, disc, size)
    return true
end

function intersect(b1::Box, b2::Box)
    return b1[RIGHT] >= b2[LEFT] &&
           b2[RIGHT] >= b1[LEFT] &&
           b1[TOP] >= b2[BOTTOM] &&
           b2[TOP] >= b1[BOTTOM]
end

function search(tree::Tree{T}, extent::Box) where T
    results = Vector{Tuple{T, Box}}()
    if tree.root === nothing
        return results
    end
    
    stack = Tuple{Node{T}, Int, Int}[]
    push!(stack, (tree.root, 1, 0))
    
    while !isempty(stack)
        node, disc, state = pop!(stack)
        hort = ((disc - 1) & 1) + 1
        
        if state == 0
            push!(stack, (node, disc, 1))
            if node.item !== nothing && intersect(extent, node.size)
                push!(results, (node.item, node.size))
            end
        elseif state == 1
            push!(stack, (node, disc, 2))
            if node.sons[1] !== nothing
                should_push = false
                if ((disc - 1) & 2) != 0
                    if extent[hort] <= node.size[disc] && extent[hort+2] >= node.lo_min_bound
                        should_push = true
                    end
                else
                    if extent[hort] <= node.other_bound && extent[hort+2] >= node.lo_min_bound
                        should_push = true
                    end
                end
                if should_push
                    push!(stack, (node.sons[1], next_disc(disc), 0))
                end
            end
        elseif state == 2
            if node.sons[2] !== nothing
                should_push = false
                if ((disc - 1) & 2) != 0
                    if extent[hort] <= node.hi_max_bound && extent[hort+2] >= node.other_bound
                        should_push = true
                    end
                else
                    if extent[hort] <= node.hi_max_bound && extent[hort+2] >= node.size[disc]
                        should_push = true
                    end
                end
                if should_push
                    push!(stack, (node.sons[2], next_disc(disc), 0))
                end
            end
        end
    end
    return results
end

function hard_delete!(tree::Tree{T}, item::T, size::Box) where T
    initial_count = tree.item_count
    tree.root = hard_delete_recursive!(tree.root, 1, item, size, tree)
    return tree.item_count < initial_count
end

function hard_delete_recursive!(node::Union{Node{T}, Nothing}, disc::Int, item::T, size::Box, tree::Tree{T}) where T
    if node === nothing
        return nothing
    end

    if node.item !== nothing && node.item == item
        if node.sons[1] === nothing && node.sons[2] === nothing
            tree.item_count -= 1
            return nothing
        end

        if node.sons[2] !== nothing
            q_item, q_size = find_extreme(node.sons[2], next_disc(disc), disc, true)
            node.item = q_item
            node.size = q_size
            node.sons[2] = hard_delete_recursive!(node.sons[2], next_disc(disc), q_item, q_size, tree)
        else
            q_item, q_size = find_extreme(node.sons[1], next_disc(disc), disc, false)
            node.item = q_item
            node.size = q_size
            node.sons[1] = hard_delete_recursive!(node.sons[1], next_disc(disc), q_item, q_size, tree)
        end
        return node
    end

    val = size[disc] - node.size[disc]
    if val == 0
        ndisc = next_disc(disc)
        while ndisc != disc
            val = size[ndisc] - node.size[ndisc]
            if val != 0
                break
            end
            ndisc = next_disc(ndisc)
        end
        if val == 0
            val = 1
        end
    end

    child_idx = val >= 0 ? 2 : 1
    node.sons[child_idx] = hard_delete_recursive!(node.sons[child_idx], next_disc(disc), item, size, tree)
    return node
end

function find_extreme(node::Node{T}, node_disc::Int, target_disc::Int, find_min::Bool) where T
    best_item = node.item
    best_size = node.size

    search_loson = node.sons[1] !== nothing
    search_hison = node.sons[2] !== nothing

    if node_disc == target_disc
        if find_min
            search_hison = false
        else
            search_loson = false
        end
    end

    if search_loson
        l_item, l_size = find_extreme(node.sons[1], next_disc(node_disc), target_disc, find_min)
        if find_min
            if l_size[target_disc] < best_size[target_disc]
                best_size = l_size
                best_item = l_item
            end
        else
            if l_size[target_disc] > best_size[target_disc]
                best_size = l_size
                best_item = l_item
            end
        end
    end

    if search_hison
        h_item, h_size = find_extreme(node.sons[2], next_disc(node_disc), target_disc, find_min)
        if find_min
            if h_size[target_disc] < best_size[target_disc]
                best_size = h_size
                best_item = h_item
            end
        else
            if h_size[target_disc] > best_size[target_disc]
                best_size = h_size
                best_item = h_item
            end
        end
    end

    return best_item, best_size
end

function find_recursive(elem::Union{Node{T}, Nothing}, disc::Int, item::T, size::Box) where T
    if elem === nothing
        return false
    end
    if elem.item !== nothing && elem.item == item
        return true
    end

    val = size[disc] - elem.size[disc]
    if val == 0
        ndisc = next_disc(disc)
        while ndisc != disc
            val = size[ndisc] - elem.size[ndisc]
            if val != 0
                break
            end
            ndisc = next_disc(ndisc)
        end
        if val == 0
            val = 1
        end
    end

    child_idx = val >= 0 ? 2 : 1
    return find_recursive(elem.sons[child_idx], next_disc(disc), item, size)
end

is_member(tree::Tree{T}, item::T, size::Box) where T = find_recursive(tree.root, 1, item, size)
count_items(tree::Tree{T}) where T = tree.item_count - tree.dead_count

end

File: coord64/go/kd.go
package kd


// Box defines a 2D bounding box [left, bottom, right, top]
// Box defines a 2D bounding box [left, bottom, right, top]
type Box [4]int64

const (
	Left   = 0
	Bottom = 1
	Right  = 2
	Top    = 3
)

// Status represents return values for operations
type Status int

const (
	OK       Status = 1
	NoMore   Status = 2
	NotImpl  Status = -3
	NotFound Status = -4
)

// Node represents an element in the KD-tree
type Node struct {
	Item         interface{}
	Size         Box
	LoMinBound   int64
	HiMaxBound   int64
	OtherBound   int64
	Sons         [2]*Node
}

// Tree represents the KD-tree structure
type Tree struct {
	Root           *Node
	ItemCount      int
	DeadCount      int
	Extent         Box
	ItemsBalanced  int
}

// Create returns a new empty KD-tree
func Create() *Tree {
	return &Tree{}
}

// Disc returns the discriminator for a given level
func Disc(level int) int {
	return level % 4
}

// ... more implementation to follow

func (t *Tree) Insert(data interface{}, size Box) {
	if data == nil {
		panic("KD: attempt to insert nil data")
	}

	if t.Root == nil {
		t.Root = &Node{
			Item:       data,
			Size:       size,
			LoMinBound: size[0],
			HiMaxBound: size[2],
			OtherBound: size[0],
		}
		t.Extent = size
		t.ItemCount = 1
		return
	}

	if t.findItem(t.Root, 0, data, size, false) != nil {
		t.ItemCount++
		// Update tree extent
		if size[Left] < t.Extent[Left] { t.Extent[Left] = size[Left] }
		if size[Right] > t.Extent[Right] { t.Extent[Right] = size[Right] }
		if size[Top] > t.Extent[Top] { t.Extent[Top] = size[Top] }
		if size[Bottom] < t.Extent[Bottom] { t.Extent[Bottom] = size[Bottom] }
	}
}

func nextDisc(disc int) int {
	return (disc + 1) % 4
}

func (t *Tree) findItem(elem *Node, disc int, item interface{}, size Box, searchP bool) *Node {
	if item == elem.Item {
		if searchP {
			return elem
		}
		return nil // Duplicate not allowed for insert
	}

	var val int64 = size[disc] - elem.Size[disc]
	if val == 0 {
		ndisc := nextDisc(disc)
		for ndisc != disc {
			val = size[ndisc] - elem.Size[ndisc]
			if val != 0 {
				break
			}
			ndisc = nextDisc(ndisc)
		}
		if val == 0 {
			val = 1
		}
	}

	childIdx := 0
	if val >= 0 {
		childIdx = 1
	}

	if elem.Sons[childIdx] != nil {
		res := t.findItem(elem.Sons[childIdx], nextDisc(disc), item, size, searchP)
		if !searchP && res != nil {
			t.boundsUpdate(elem, disc, size)
		}
		return res
	}

	if searchP {
		return nil
	}

	// Insert here
	vert := int(nextDisc(disc) & 0x01)
	newNode := &Node{
		Item:       item,
		Size:       size,
		LoMinBound: size[vert],
		HiMaxBound: size[vert+2],
	}
	// Logic for other_bound from C code: 
	// items_elem->other_bound = ((NEXTDISC(disc)&0x2) ? size[vert] : size[vert+2]);
	if (nextDisc(disc) & 0x2) != 0 {
		newNode.OtherBound = size[vert]
	} else {
		newNode.OtherBound = size[vert+2]
	}

	elem.Sons[childIdx] = newNode
	t.boundsUpdate(elem, disc, size)
	return newNode
}

func (t *Tree) boundsUpdate(elem *Node, disc int, size Box) {
	vert := int(disc & 0x01)
	if size[vert] < elem.LoMinBound { elem.LoMinBound = size[vert] }
	if size[vert+2] > elem.HiMaxBound { elem.HiMaxBound = size[vert+2] }
	if (disc & 0x02) != 0 {
		if size[vert] < elem.OtherBound { elem.OtherBound = size[vert] }
	} else {
		if size[vert+2] > elem.OtherBound { elem.OtherBound = size[vert+2] }
	}
}

type State int

const (
	ThisOne State = 0
	LoSon   State = 1
	HiSon   State = 2
	Done    State = 3
)

type save struct {
	node  *Node
	disc  int
	state State
}

type Generator struct {
	extent Box
	stack  []save
}

func (t *Tree) Start(area Box) *Generator {
	g := &Generator{
		extent: area,
	}
	if t.Root != nil {
		g.stack = append(g.stack, save{node: t.Root, disc: 0, state: ThisOne})
	}
	return g
}

func (g *Generator) Next() (interface{}, Box, bool) {
	for len(g.stack) > 0 {
		topIdx := len(g.stack) - 1
		top := &g.stack[topIdx]
		node := top.node
		m := top.disc
		hort := m & 0x01

		switch top.state {
		case ThisOne:
			g.stack[topIdx].state = LoSon
			if node.Item != nil && Intersect(g.extent, node.Size) {
				return node.Item, node.Size, true
			}
		case LoSon:
			g.stack[topIdx].state = HiSon
			if node.Sons[0] != nil {
				shouldPush := false
				if (m & 0x02) != 0 { // RIGHT or TOP
					if g.extent[hort] <= node.Size[m] && g.extent[hort+2] >= node.LoMinBound {
						shouldPush = true
					}
				} else { // LEFT or BOTTOM
					if g.extent[hort] <= node.OtherBound && g.extent[hort+2] >= node.LoMinBound {
						shouldPush = true
					}
				}
				if shouldPush {
					g.stack = append(g.stack, save{node: node.Sons[0], disc: nextDisc(m), state: ThisOne})
					continue
				}
			}
		case HiSon:
			g.stack[topIdx].state = Done
			if node.Sons[1] != nil {
				shouldPush := false
				if (m & 0x02) != 0 { // RIGHT or TOP
					if g.extent[hort] <= node.HiMaxBound && g.extent[hort+2] >= node.OtherBound {
						shouldPush = true
					}
				} else { // LEFT or BOTTOM
					if g.extent[hort] <= node.HiMaxBound && g.extent[hort+2] >= node.Size[m] {
						shouldPush = true
					}
				}
				if shouldPush {
					g.stack = append(g.stack, save{node: node.Sons[1], disc: nextDisc(m), state: ThisOne})
					continue
				}
			}
		case Done:
			g.stack = g.stack[:topIdx]
		}
	}
	return nil, Box{}, false
}

func Intersect(b1, b2 Box) bool {
	return b1[Right] >= b2[Left] &&
		b2[Right] >= b1[Left] &&
		b1[Top] >= b2[Bottom] &&
		b2[Top] >= b1[Bottom]
}

// IsMember checks if an item exists in the tree
func (t *Tree) IsMember(data interface{}, size Box) bool {
	return t.findItem(t.Root, 0, data, size, true) != nil
}

// Count returns the number of active items in the tree
func (t *Tree) Count() int {
	return t.ItemCount - t.DeadCount
}

// Delete marks an item as dead (soft delete)
func (t *Tree) Delete(data interface{}, size Box) bool {
	node := t.findItem(t.Root, 0, data, size, true)
	if node != nil && node.Item != nil {
		node.Item = nil
		t.DeadCount++
		return true
	}
	return false
}

// HardDelete physically removes an item from the tree and restructures it
func (t *Tree) HardDelete(item interface{}, size Box) bool {
	initialCount := t.ItemCount
	t.Root = t.hardDeleteRecursive(t.Root, 0, item, size)
	return t.ItemCount < initialCount
}

func (t *Tree) hardDeleteRecursive(node *Node, disc int, item interface{}, size Box) *Node {
	if node == nil {
		return nil
	}

	if node.Item == item {
		if node.Sons[0] == nil && node.Sons[1] == nil {
			t.ItemCount--
			return nil
		}
		if node.Sons[1] != nil { // HISON exists, find MIN in target disc
			qItem, qSize := t.findExtreme(node.Sons[1], nextDisc(disc), disc, true)
			node.Item = qItem
			node.Size = qSize
			node.Sons[1] = t.hardDeleteRecursive(node.Sons[1], nextDisc(disc), qItem, qSize)
		} else { // LOSON exists, find MAX in target disc
			qItem, qSize := t.findExtreme(node.Sons[0], nextDisc(disc), disc, false)
			node.Item = qItem
			node.Size = qSize
			node.Sons[0] = t.hardDeleteRecursive(node.Sons[0], nextDisc(disc), qItem, qSize)
		}
		return node
	}

	val := size[disc] - node.Size[disc]
	if val == 0 {
		ndisc := nextDisc(disc)
		for ndisc != disc {
			val = size[ndisc] - node.Size[ndisc]
			if val != 0 {
				break
			}
			ndisc = nextDisc(ndisc)
		}
		if val == 0 {
			val = 1
		}
	}

	childIdx := 0
	if val >= 0 {
		childIdx = 1
	}

	node.Sons[childIdx] = t.hardDeleteRecursive(node.Sons[childIdx], nextDisc(disc), item, size)
	return node
}

func (t *Tree) findExtreme(node *Node, nodeDisc int, targetDisc int, findMin bool) (interface{}, Box) {
	bestItem := node.Item
	bestSize := node.Size

	searchLoson := node.Sons[0] != nil
	searchHison := node.Sons[1] != nil

	if nodeDisc == targetDisc {
		if findMin {
			searchHison = false // HISON elements are >= node, cannot be MIN
		} else {
			searchLoson = false // LOSON elements are < node, cannot be MAX
		}
	}

	if searchLoson {
		lItem, lSize := t.findExtreme(node.Sons[0], nextDisc(nodeDisc), targetDisc, findMin)
		if findMin {
			if lSize[targetDisc] < bestSize[targetDisc] {
				bestSize = lSize
				bestItem = lItem
			}
		} else {
			if lSize[targetDisc] > bestSize[targetDisc] {
				bestSize = lSize
				bestItem = lItem
			}
		}
	}

	if searchHison {
		hItem, hSize := t.findExtreme(node.Sons[1], nextDisc(nodeDisc), targetDisc, findMin)
		if findMin {
			if hSize[targetDisc] < bestSize[targetDisc] {
				bestSize = hSize
				bestItem = hItem
			}
		} else {
			if hSize[targetDisc] > bestSize[targetDisc] {
				bestSize = hSize
				bestItem = hItem
			}
		}
	}

	return bestItem, bestSize
}
